pow/ethash.go

```diff
@@ -0,0 +1,572 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package pow
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+	"unsafe"
+
+	mmap "github.com/edsrzf/mmap-go"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	metrics "github.com/rcrowley/go-metrics"
+)
+
+var (
+	ErrNonceOutOfRange   = errors.New("nonce out of range")
+	ErrInvalidDifficulty = errors.New("non-positive difficulty")
+	ErrInvalidMixDigest  = errors.New("invalid mix digest")
+	ErrInvalidPoW        = errors.New("pow difficulty invalid")
+)
+
+var (
+	// maxUint256 is a big integer representing 2^256-1
+	maxUint256 = new(big.Int).Exp(big.NewInt(2), big.NewInt(256), big.NewInt(0))
+
+	// sharedEthash is a full instance that can be shared between multiple users.
+	sharedEthash = NewFullEthash("", 3, 0, "", 1, 0)
+
+	// algorithmRevision is the data structure version used for file naming.
+	algorithmRevision = 23
+
+	// dumpMagic is a dataset dump header to sanity check a data dump.
+	dumpMagic = hexutil.MustDecode("0xfee1deadbaddcafe")
+)
+
+// isLittleEndian returns whether the local system is running in little or big
+// endian byte order.
+func isLittleEndian() bool {
+	n := uint32(0x01020304)
+	return *(*byte)(unsafe.Pointer(&n)) == 0x04
+}
+
+// memoryMap tries to memory map a file of uint32s for read only access.
+func memoryMap(path string, write bool) (*os.File, mmap.MMap, []uint32, error) {
+	// Try to open the file and memory map it
+	flag, mapflag := os.O_RDONLY, mmap.RDONLY
+	if write {
+		flag, mapflag = os.O_RDWR, mmap.RDWR
+	}
+	file, err := os.OpenFile(path, flag, 0644)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	mem, err := mmap.Map(file, mapflag, 0)
+	if err != nil {
+		file.Close()
+		return nil, nil, nil, err
+	}
+	// Yay, we managed to memory map the file, here be dragons
+	header := *(*reflect.SliceHeader)(unsafe.Pointer(&mem))
+	header.Len /= 4
+	header.Cap /= 4
+
+	return file, mem, *(*[]uint32)(unsafe.Pointer(&header)), nil
+}
+
+// memoryMapAndGenerate tries to memory map a temporary file of uint32s for write
+// access, fill it with the data from a generator and then move it into the final
+// path requested.
+func memoryMapAndGenerate(path string, size uint64, generator func(buffer []uint32)) (*os.File, mmap.MMap, []uint32, error) {
+	// Ensure the data folder exists
+	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
+		return nil, nil, nil, err
+	}
+	// Create a huge temporary empty file to fill with data
+	temp := path + "." + strconv.Itoa(rand.Int())
+
+	dump, err := os.Create(temp)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	if err = dump.Truncate(int64(size)); err != nil {
+		return nil, nil, nil, err
+	}
+	dump.Close()
```

Comment:
Now that I look at this again, it looks weird that you're closing the file and reopen it just below. You can reuse the file that's already open.